pykern/fconf.py

```diff
@@ -0,0 +1,517 @@
+
+Note that the arguments were not quoted here, because they are integers.
+
+The typical usage is::
+
+    fconf.parse_all(some_dir)
+
+where `some_dir` is a `py.path` that contains python and yaml files to parse.
+You can also call `Parser` directly with the a list of files to parse.
+
+As noted, the arguments to a macro are Python. Actually, the entire
+value is a Python expression, and the only requirement is that it
+begin with a Python word, which can be any Python function. For example::
+
+    count_down:
+      - reversed(range(4))
+
+Produces the following Python Structure::
+
+    {'count_down': [3, 2, 1, 0]}
+
+This allows for a lot of possibilities within the YAML alone.
+
+The Python and YAML macros can call each other. Inside a Python file,
+just use self to call something::
+
+    def count_up(self, top):
+        return reversed(self.count_down(top))
+
+Note that the above count_down example used a special feature of
+result merging. When a Macro returns a list (generator, tuple, etc.)
+in a list context, it will merge with the list. Strings are used in
+place. For example::
+
+    mixed:
+      - range(3)
+      - make_uri('sirepo.com')
+
+Produces::
+
+    {'mixed': [0, 1, 2, 'https://sirepo.com']}
+
+The same is true of dicts. For example::
+
+    fconf_macros:
+      base(num):
+        v${num}.radia.run:
+          description: VM ${num}
+
+    all_hosts:
+      base(3):
+      base(5):
+
+Produces::
+
+    {
+        "all_hosts": {
+            "v3.radia.run": {
+                "description": "VM 3",
+            },
+            "v5.radia.run": {
+                "description": "VM 5",
+            },
+        },
+    }
+
+There are some other useful features. You can refer to any previous
+declared value using variable expansion. For example::
+
+    a:
+      b:
+        c: 3
+    d: ${a}
+    e: ${a.b.c}
+
+
+Produces::
+
+    {
+        'a': {'b': {'c': 3}},
+        'd': {'b': {'c': 3}},
+        'e': 3,
+    }
+
+This allows for flexible constants. Those global strings can be used
+in YAML Macros.
+
+YAML files are evaluated before they are merged. However, they are all
+merged before the next file is evaluated. This allows a main
+"constants" file, for example, to direct the flow of the subsequent files.
+
+:copyright: Copyright (c) 2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+
+"""
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdc, pkdlog, pkdp
+import contextlib
+import copy
+import inspect
+import pykern.pkrunpy
+import re
+import pykern.pkyaml
+import pykern.pkcollections
+import pykern.pkio
+
+
+#: parse_files macro expansion pattern
+_TEXT_MACRO_NAME = re.compile(
+    r"^([a-z]\w*)\(((?:\s*[a-z]\w*\s*)?(?:,\s*[a-z]\w*\s*)*)\)$", flags=re.IGNORECASE
+)
+
+_MACRO_CALL = re.compile(r"^([a-z]\w*)\((.*)\)$", flags=re.IGNORECASE + re.DOTALL)
+
+_ARG_SEP = re.compile(r"\s*,\s*|\s+")
+
+_FVAR = re.compile(r"\$\{(\S+)\}")
```

Comment:
I think something more descriptive than "F" would be helpful.

Reply:
I can't think of anything. It's an internal name (for now). It's used for both argument and "current value in the eval" expansions. It's used in the code. I tried to clear up with a comment 5ee9801